### PR TITLE
Element id interface for material

### LINKED ISSUE #16005: Some ElementIDInterface functions need to be override in Material
## Reason
`Material` is different from `DGKernel` or `InterfaceMaterial`, where we can retrieve element IDs for both current and neighboring elements. `Material` owns three copies: element, face, neighbor. The first two should always get the current element ID while the last one gets the neighboring element ID. If we do not take care of this in `Material`, user materials much do the `if (_neighbor)` check when getting element IDs.

## Design
A better design that can avoid potential bugs.

## Impact
It can either fix potential bugs for applications or temporarily break applications.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #16005.

### CODE REVIEW

File: test/tests/auxkernels/mesh_integer/tests [hunk 1]
```diff
@@ -1,10 +1,17 @@
...
     input = 'mesh_integer.i'
     exodiff = 'mesh_integer_out.e'
+    issues = '#13764'
+    requirement = "MOOSE shall include the ability to use extra element integers."
+  [../]
+  [./dg_mesh_integer]
+    type = 'Exodiff'
+    input = 'dg_mesh_integer.i'
+    exodiff = 'dg_mesh_integer_out.e'
+    issues = '#16005'
```
Comment (YaqiWang) [inline test/tests/auxkernels/mesh_integer/tests:14]: Note: this test will generate wrong results if we do not have this fix.

**Comment by YaqiWang** [comment]
This is ready for review. @lindsayad You are probably the right one for reviewing this. It is a small PR.

File: test/tests/auxkernels/mesh_integer/tests [hunk 2]
```diff
@@ -1,10 +1,17 @@
...
     requirement = "MOOSE shall include the ability to use extra element integers."
```
Comment (lindsayad) [inline test/tests/auxkernels/mesh_integer/tests:15]: These are two different tests, so the requirements should be different

File: framework/include/materials/Material.h [hunk 3]
```diff
@@ -30,6 +30,53 @@ class Material : public MaterialBase, public Coupleable, public MaterialProperty
 
   Material(const InputParameters & parameters);
 
+  /**
+   * Gets an element integer for the proper current element with a parameter of
+   * the object derived from this interface
+   * Note: This overrides the function in ElementIDInterface to assure derived materials
+   *       to call the functions in ElementIDInterface properly.
```
Comment (lindsayad) [inline framework/include/materials/Material.h:37]: delete `to`

File: framework/include/materials/Material.h [hunk 4]
```diff
@@ -30,6 +30,53 @@ class Material : public MaterialBase, public Coupleable, public MaterialProperty
...
+  }
+  /**
+   * Directly calling this function is not needed for materials because the same material has
+   * three copies for element interior, element side and neighbor side.
+   */
+  virtual const dof_id_type & getElementIDNeighbor(const std::string & id_parameter_name,
+                                                   unsigned int comp = 0) const override
+  {
+    mooseError("Directly calling getElementIDNeighbor is not allowed for materials intead of "
+               "calling getElementID");
```
Comment (lindsayad) [inline framework/include/materials/Material.h:53]: "Directly calling 'getElementIDNeighbor' is not allowed for materials. Please call 'getElementID' instead."

File: framework/include/materials/Material.h [hunk 5]
```diff
@@ -30,6 +30,53 @@ class Material : public MaterialBase, public Coupleable, public MaterialProperty
...
+  }
+  /**
+   * Directly calling this function is not needed for materials because the same material has
+   * three copies for element interior, element side and neighbor side.
+   */
+  virtual const dof_id_type &
+  getElementIDNeighborByName(const std::string & id_parameter_name) const override
+  {
+    mooseError("Directly calling getElementIDNeighborByName is not allowed for materials intead of "
+               "calling getElementIDByName");
```
Comment (lindsayad) [inline framework/include/materials/Material.h:76]: "Directly calling 'getElementIDNeighborByName' is not allowed for materials. Please call 'getElementIDByName' instead."

**Review by lindsayad — CHANGES_REQUESTED** [review]
Looks pretty good to me. Just the small requested changes

**Comment by YaqiWang** [comment]
Comments are addressed.

**Review by lindsayad — APPROVED** [review]
Thank you for the additional documentation as well!

**Comment by YaqiWang** [comment]
Thank you @lindsayad !

**Pull request merged**